src/phpDocumentor/FilesystemFactory.php

```diff
@@ -0,0 +1,89 @@
+<?php
+/**
+ * This file is part of phpDocumentor.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @copyright 2010-2015 Mike van Riel<mike@phpdoc.org>
+ * @license   http://www.opensource.org/licenses/mit-license.php MIT
+ * @link      http://phpdoc.org
+ */
+
+namespace phpDocumentor;
+
+use \LogicException;
+use League\Flysystem\Adapter\Local;
+use League\Flysystem\Filesystem;
+use League\Flysystem\MountManager;
+use \DI\Container;
+
+/**
+ * Class FilesystemFactory
+ */
+final class FilesystemFactory
+{
+    /** @var MountManager */
+    private $mountManager;
+
+    /** @var Container */
+    private $container;
+
+    /** @var Filesystem */
+    private $filesystem;
+
+    /**
+     * @param MountManager $mountManager
+     * @param Container $container
+     */
+    public function __construct(MountManager $mountManager, Container $container)
+    {
+        $this->mountManager = $mountManager;
+        $this->container = $container;
+    }
+
+    /**
+     * Returns a Filesystem instance based on the scheme of the provided Dsn
+     *
+     * @param Dsn $dsn
+     * @return Filesystem
+     */
+    public function create(Dsn $dsn)
+    {
+        $dsnId = spl_object_hash($dsn);
+
+        if (! $this->setFilesystemWhenCached($dsnId)) {
+            if ($dsn->getScheme() === 'file') {
+
+                $path = $dsn->getPath();
+                $this->filesystem = new Filesystem(new Local(__DIR__. "/" . $path));
+
+            } else {
+                //This will be implemented as soon as the CloneRemoteGitToLocal adapter is finished
+                throw new \InvalidArgumentException('http and https are not supported yet');
+            }
+
+            $this->mountManager->mountFilesystem($dsnId, $this->filesystem);
+        }
+
+
+        return $this->filesystem;
+    }
+
+    /**
+     * If the requested filesystem is available in MountManager
+     * set the filesystem property to this filesystem
+     *
+     * @param string $dsnId
+     * @return bool
+     */
+    private function setFilesystemWhenCached($dsnId)
+    {
+        try {
+            $this->filesystem = $this->mountManager->getFilesystem($dsnId);
+            return true;
+        } catch (LogicException $e) {
+            echo 'Caught exception: ',  $e->getMessage(), "\n";
```

Comment:
Please do not use `echo` (or otherwise directly write to the STDOUT) in a code base when you want to display debug data; most codebases have logging mechanisms to deal with this, and if they don't (like phpDocumentor 3) you can use `trigger_error`.

In this case you can wonder why an exception occurs and what happens now. If no filesystem property was present before than this class will have a fatal error once it hits line 59 because the filesystem property is empty; or worse: if a previous call to create was successful and this wasn't than the wrong file system object is returned because it will return the previous call's filesystem.

> This is actually a really good reason why you don't want to store the filesystem in a property but keep the filesystem local to the create method, if you return the property in this method instead of a boolean then you don't need the property at all.